### PR TITLE
[ie/beacon] Add extractor for Beacon.tv

### PR DESCRIPTION
**Pull request by Deukhoofd**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add an extractor to support downloading from https://beacon.tv, a new media website launched by Critical Role yesterday.

This website places a JSON in an HTML tag with a specific ID and type. This tag is extracted, the JSON is parsed, and the m3u8 URL and metadata are retrieved from the JSON.

Authentication on the site is done through cookies. This means that downloading videos that require authentication currently depends on the --cookies-from-browser command line flag. Videos that are public can be downloaded without this flag.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by Deukhoofd** [comment]
Abandoning, as I just realized there's a way cleaner method to do what it needs to do.

**Comment by bashonly** [comment]
You can push to the same branch and repurpose this PR

**Comment by Deukhoofd** [comment]
Nevermind the earlier message. I tried to use the Beacon API at https://beacon.tv/api/graphql. This works generally, and would simplify the flow of the code a lot. Unfortunately, for authenticated requests this requires an authentication token. This token is not retrievable from cookies, and requires a fairly complicated Oauth flow (using Memberful, a Wordpress plugin). This flow involves captions and other annoyances I wasn't able to work around.

While the direct API handling would have been a lot nicer, scraping the information from the HTML also functions. As such this should be fine for now.

**Comment by Meldanor** [comment]
I get some errors for their "Cooldown" series. Here is an example output:

```
[debug] Command-line config: ['--cookies', 'cookies.txt', 'https://beacon.tv/content/3-90-cr-cooldown-c3-e90', '-v']
[debug] User config "config": ['--add-chapters', '--sub-langs', 'en.*,de', '--embed-subs', '-o', '%(playlist_index&{} |)s%(title)s.%(ext)s']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.04.09 from yt-dlp/yt-dlp [ff0779267] (zip)
[debug] Python 3.11.2 (CPython aarch64 64bit) - Linux-6.6.28+rpt-rpi-2712-aarch64-with-glibc2.36 (OpenSSL 3.0.11 19 Sep 2023, glibc 2.36)
[debug] exe versions: ffmpeg 5.1.4-0 (setts), ffprobe 5.1.4-0
[debug] Optional libraries: Cryptodome-3.11.0, certifi-2022.09.24, requests-2.28.1, sqlite3-3.40.1, urllib3-1.26.12
[debug] Proxy map: {}
[debug] Request Handlers: urllib
[debug] Loaded 1811 extractors
[BeaconTv] Extracting URL: https://beacon.tv/content/3-90-cr-cooldown-c3-e90
[BeaconTv] 3-90-cr-cooldown-c3-e90: Downloading webpage
ERROR: 3-90-cr-cooldown-c3-e90: An extractor error has occurred. (caused by KeyError('language')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/common.py", line 734, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/beacon.py", line 80, in _real_extract
    language = track['language'].lower()
               ~~~~~^^^^^^^^^^^^
KeyError: 'language'
```

**Comment by Deukhoofd** [comment]
Thanks for the heads up, should be fixed now! If Beacon does not explicitly set the language on a subtitle, we now default it to English.

**Comment by Meldanor** [comment]
Great, I've tested it and it is working now, thank you for the work and the quick response!

**Comment by cyplo** [comment]
awesome work :) is there something missing before this can be merged btw ? would love to use it via yt-dlp packaged with my distro :)

**Comment by Deukhoofd** [comment]
I'm still waiting for a PR review unfortunately

File: yt_dlp/extractor/beacon.py
Comment (seproDev) [inline yt_dlp/extractor/beacon.py:47]: How about we just use the jwplayer data?
```suggestion

        return self._parse_jwplayer_data(traverse_obj(
            content_data, ('contentVideo', 'video', 'videoData', {json.loads})))
```

If you want to include some additional data from the page api you should merge the two info dicts rather than re-implementing the jwplayer code.

**Review by seproDev — CHANGES_REQUESTED** [review]
Sorry it took so long for someone to take a proper look at this.

**Comment by Deukhoofd** [comment]
Awesome, thanks for the review, didn't know some of these great helper functions existed!

Should have integrated the suggested changes now. Quickly tested with a couple videos, and everything appears to be still in order.

Thanks!

File: yt_dlp/extractor/beacon.py
Comment (seproDev) [inline yt_dlp/extractor/beacon.py:46]: I would inline this and also validate the returned data type
```suggestion
        return {
            **self._parse_jwplayer_data(traverse_obj(
                content_data, ('contentVideo', 'video', 'videoData', {json.loads})), video_id),
            **traverse_obj(content_data, {
                'title': ('title', {str}),
                'description': ('description', {str}),
                'timestamp': ('publishedAt', {parse_iso8601}),
            }),
        }
```

**Comment by Deukhoofd** [comment]
Thanks, not super at home at Python, but that is a lot cleaner!

**Review by seproDev — APPROVED** [review]

**Review by bashonly — APPROVED** [review]

**Comment by seproDev** [comment]
@Deukhoofd I've pushed a slight refactor, could you please verify that everything still works as expected?

**Comment by Deukhoofd** [comment]
Everything appears to still be functioning

**Pull request merged**
